lxd/db/instances.go

```diff
@@ -321,64 +321,78 @@ SELECT instances.name, nodes.id, nodes.address, nodes.heartbeat
+		if err != nil {
+			return errors.Wrap(err, "Load instances")
+		}
 
-	// Map to check which projects have the profiles features on.
-	projectHasProfiles := map[string]bool{}
-	for _, project := range projects {
-		projectHasProfiles[project.Name] = shared.IsTrue(project.Config["features.profiles"])
-	}
+		projects, err := tx.GetProjects(ProjectFilter{})
+		if err != nil {
+			return errors.Wrap(err, "Load projects")
+		}
 
-	profiles, err := c.GetProfiles(ProfileFilter{})
-	if err != nil {
-		return nil, errors.Wrap(err, "Load profiles")
-	}
+		// Index of all projects by name and record which projects have the profiles feature.
+		for i, project := range projects {
+			projectMap[project.Name] = projects[i]
+			projectHasProfiles[project.Name] = shared.IsTrue(project.Config["features.profiles"])
+		}
 
-	// Index of all profiles by project and name.
-	profilesByProjectAndName := map[string]map[string]Profile{}
-	for _, profile := range profiles {
-		profilesByName, ok := profilesByProjectAndName[profile.Project]
-		if !ok {
-			profilesByName = map[string]Profile{}
-			profilesByProjectAndName[profile.Project] = profilesByName
+		profiles, err := tx.GetProfiles(ProfileFilter{})
+		if err != nil {
+			return errors.Wrap(err, "Load profiles")
 		}
-		profilesByName[profile.Name] = profile
+
+		// Index of all profiles by project and name.
+		for _, profile := range profiles {
+			profilesByName, ok := profilesByProjectAndName[profile.Project]
+			if !ok {
+				profilesByName = map[string]Profile{}
+				profilesByProjectAndName[profile.Project] = profilesByName
+			}
+			profilesByName[profile.Name] = profile
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
 	}
 
-	for i, instance := range instances {
+	// Call the instanceFunc provided for each instance after the transaction has ended, as we don't know if
+	// the instanceFunc will be slow or may need to make additional DB queries.
+	for _, instance := range instances {
 		profiles := make([]api.Profile, len(instance.Profiles))
 
+		// If the instance's project does not have the profiles feature enabled,
+		// we fall back to the default project.
 		profilesProject := instance.Project
-
-		// If the instance's project does not have the profiles feature
-		// enable, we fall back to the default project.
 		if !projectHasProfiles[profilesProject] {
-			profilesProject = "default"
+			profilesProject = "default" // Equivalent to project.Default constant.
 		}
 
 		for j, name := range instance.Profiles {
 			profile := profilesByProjectAndName[profilesProject][name]
 			profiles[j] = *ProfileToAPI(&profile)
 		}
 
-		instances[i].Config = ExpandInstanceConfig(instance.Config, profiles)
-		instances[i].Devices = ExpandInstanceDevices(deviceConfig.NewDevices(instance.Devices), profiles).CloneNative()
+		err = instanceFunc(instance, projectMap[instance.Project], profiles)
+		if err != nil {
+			return err
+		}
 	}
 
-	return instances, nil
+	return err
```

Comment:
What's `err` at that point?

Maybe I'm mis-reading the diff, but it looks like all cases of err != nil are directly handled so this should just be `return nil`?

Reply:
@stgraber ah yes thats left over from the earlier form of the function, will change to `return nil`.